In [5]:
from math import factorial

### Probability Interview Questions

#### 5.1 Google (easy)

Two teams play a series of games (best of 7 -- whoenever wins 4 games first)  
in which each team has a 50% chance of winning any given round (no draws allowed).

What is the probability that the series goes to 7 games?

--- 

In [65]:
def choose(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

For the game to go to 7 games, each team must have won exactly 3 games each in the first 6.

In [62]:
choose(6, 3)

20.0

Divided by the total number of possible 6 games series.

In [63]:
choose(6, 3) / 2**6 # or 20 / 64 --> 5 / 16

0.3125

#### 5.2 JP Morgan (easy)

Say you roll a die three times. What is the probability of getting 2 sixes in a row?

--- 

In [21]:
p_of_rolling_six = 1 / 6
p_of_rolling_other = 5 / 6

Given by the sum of the following probabilities:
- roll two sixes, roll any other number
- roll any other number, roll two sixes
- roll three sixes in a row

In [26]:
round((p_of_rolling_six ** 2) * p_of_rolling_any_other_number * 2 + p_of_rolling_six ** 3, 4)

0.0509

#### 5.3 Uber (easy)

You roll three dice, one after another.

What is the probability that you obtain three numbers in a strictly increasing order?

--- 

All possible ways to throw 3 dice such that the numbers are strictly increasing:
* 1, 2, 3-6
* 1, 3, 4-6
* 1, 4, 5-6
* 1, 5, 6
* 2, 3, 4-6
* 2, 4, 5-6
* 2, 5, 6
* 3, 4, 5-6
* 3, 5, 6
* 4, 5, 6

In [33]:
def permutations(n, k):
    return factorial(n) / (factorial(n - k))

Probability is given by

In [49]:
((1/6) ** 3) * 4 + (1/6)**2 * (4/6) + ((1/6)**2 * (3/6)) * 2 + ((1/6)**2 * 2/6) * 3

0.09259259259259259

Another way of solving it is, getting the probability that the numbers are different

In [50]:
1 * (5/6) * (4/6)  # 5/9

0.5555555555555556

Conditioned on there being three different numbers, there is exactly one particular   
sequence that will be in a strictly increasing order, and this sequence occurs with   
probability `1/3! = 1/6`.

And so the probability is given by `5/9 * 1/6 = 5/54`

#### 5.4 Zenefits (easy)

Assume you have a deck of 100 cards with values ranging from 1 to 100, and that you draw two cards at random without replacement.

What is the probability that the number of one card is precisely double that of the other?

--- 

In [60]:
((1/50) * (1/99)) * 50

0.010101010101010102

Another way of solving it is:
    
There are a total of `100 choose 2 = 4950` ways to choose two cards at random from 100.  
There are exactly 50 pairs that satisfy the condition: (1,2) .. (50, 100).  
Therefore the probability is: `50 / 4950 ~= 0.01`

#### 5.5 JP Morgan (easy)

Imagine you are in a 3D space. From (0,0,0) to (3,3,3), how many paths are there if you can move only up, right and forward?

--- 

Getting to (3,3,3) requires 9 moves. Using these 9 moves, it must be the case that there are exactly  
three moves in each of the three directions (up, right, and forward). There are therefore 9! ways to  
order the 9 moves in any given direction. We must divide by 3! for each direction to avoid overcounting  
since each up move is indistinguishable. Therefore, the number of paths is: `9! / (3!3!3!) = 1680`.

#### 5.6 Amazon (easy)

One in a thousand people have a particular disease, and the test for the disease is 98%  
correct in testing for the disease. On the other hand, the test has 1% error rate if the  
person being tested does not have the disease.

If someone tests positive, what are the odds that they have the disease?

--- 

In [69]:
def bayes_rule(pa, pb, pba):
    return (pba * pa) / pb

The probability that a person has the disease, given that they tested positive is

```
P(A|B) = (P(B|A) * P(A)) / P(B)
```

where A is the event of a person having the disease and B is the event that this person tests positive.

In [85]:
# probability of having the disease
pa = 1 / 1000

# probability of testing positive
pb = (1/1000) * 0.98 + (999/1000) * 0.01

# probability of testing positive given that a person has the disease
pba = 0.98

In [79]:
bayes_rule(pa, pb, pba)

0.08933454876937101

Therefore the probability of someone having the disease given that they tested positive is 8.93%.

#### 5.7 Facebook (easy)

Assume two coins, one fair (having one side heads and one side tails) and the other unfair (having both sides tails).  
You pick one at random, flip it five times, and observe that it comes up as tails all five times.

What is the probability that you are flipping the unfair coin?

--- 

We can use Bayes Rule again, where we denote A as the event of picking the unfair coin  
and B as the event of flipping 5 tails, then we want `P(A|B)`.

In [109]:
# probability of picking the unfair coin
pa = 0.5

# probability of flipping 5 tails
pb = (0.5 * 0.5 ** 5) + 0.5 * 1

# probability of flipping 5 tails given that we picked the unfair coin is
pba = 1

In [110]:
bayes_rule(pa, pb, pba)

0.9696969696969697

#### 5.8 Goldman Sachs (easy)

Players A and B are playing a game where they take turns flipping a biased coin, with p probability  
of landing on heads (and winning). Player A starts the game, and then the players pass the coin back  
and forth until one person flips heads and wins.

What is the probability that A wins?

--- 

Let P(A) be the probability that A wins. Then, we know the following to be true:

1. If A flips heads initially, A wins with probability 1.
2. If A flips tails initially, and then B flips a tail, then it is as if neither flip had occurred, and so A wins with probability P(A).

Combining the two outcomes, we have: P(A) = p + (1 - p)^2P(A), and simplifying this yields P(A) = p + P(A) - 2pP(A) + p^2P(A) so that p^2P(A) - 2pP(A) + p = 0
and hence:

$$
P(A) = \frac{1}{2 - p}
$$

#### 5.9 Microsoft (easy)

Three friends in Seattle each told you it is rainy, and each person has a 1/3 probability of lying.

What is the probability that Seattle is rainy, assuming that the likelihood of rain on any given day is 0.25?

--- 

Using Bayes, trying to find P(R|F), namely the probability that it is raining given that your three friends said so.

P(R) = 0.25  
P(F) = 0.25 * (2/3)**3 + 0.75 * (1/3)**3  
P(F|R) = (2/3)**3

So, we have 

$$
P(R|F) = \frac{P(R)P(F|R)}{P(F)} = \frac{0.25 \times (2/3)^3}{0.25 \times (2/3)^3 + 0.75 \times (1/3)^3 } = 0.7272
$$

#### 5.10 Bloomberg (easy)

You draw a circle and choose two chords at random. What is the probability that those chords will intersect?

---  

By definition, a chord is a line segment where the two endpoints lie on the circle. Therefore, two  
arbitrary chords can always be represented by any four points chosen on the circle. If you choose to  
represent the first chord by two of teh four points, then you have:

In [231]:
choose(4,2)

6

choices of choosing the two points to represent chord 1. However, note that in this counting, we are  
duplicating the count of each chord twice, since a chord with endpoints p1 and p2 is the same as a chord  
with endpoints p2 and p1. Therefore, the proper number of valid chords is:

In [232]:
1/2 * choose(4,2)

3.0

Among these three configurations, in only one of them will the chords intersect, so the desired probability is p = 1/3

#### 5.11 Morgan Stanley (easy)

You and your friend are playing a game. The two of you will continue to toss a coin until the  
sequence HH or TH shows up. If HH shows up first, you win. If TH shows up first, your friend wins.

What is the probability of you winning?

---  

All the different options



In [280]:
from itertools import product

In [320]:
tosses = list(product('HT', repeat=4))

In [325]:
tosses

[('H', 'H', 'H', 'H'),
 ('H', 'H', 'H', 'T'),
 ('H', 'H', 'T', 'H'),
 ('H', 'H', 'T', 'T'),
 ('H', 'T', 'H', 'H'),
 ('H', 'T', 'H', 'T'),
 ('H', 'T', 'T', 'H'),
 ('H', 'T', 'T', 'T'),
 ('T', 'H', 'H', 'H'),
 ('T', 'H', 'H', 'T'),
 ('T', 'H', 'T', 'H'),
 ('T', 'H', 'T', 'T'),
 ('T', 'T', 'H', 'H'),
 ('T', 'T', 'H', 'T'),
 ('T', 'T', 'T', 'H'),
 ('T', 'T', 'T', 'T')]

In [340]:
you, friend = 0, 0

for toss in tosses:
    for i in range(2, 5):
        if toss[i - 2: i] == ('H', 'H'):
            you += 1
            break
        elif toss[i - 2: i] == ('T', 'H'):
            friend += 1
            break
            
print('The probability of you winning is {} or 1/3.5'.format(round(you / (you + friend), 4)))

The probability of you winning is 0.2857 or 1/3.5


Simple trick that simplifies the problem is as following. Note that, if T is ever flipped, you cannot  
then reach HH before your friend reaches TH, since the first heads thereafter will result in them winning.  
Therefore, the probability of you winning is limited to just flipping an HH initially, which we know is  
given by the following probability

$$
P(HH) = \frac{1}{2}\times \frac{1}{2} = \frac{1}{4}
$$